nbclient/client.py

```diff
@@ -426,6 +483,7 @@ async def async_start_new_kernel_client(self) -> KernelClient:
             await self._async_cleanup_kernel()
             raise
         self.kc.allow_stdin = False
+        await run_hook(self.on_execution_start)
```

Comment:
For symmetry's sake, why not also include two extra notebook-level hooks: `on_notebook_complete` and `on_notebook_error`, that map on to the two cell execution hooks but at a notebook level?

That way you would have the same basic pattern for both the notebook, and for each cell in the notebook:

- hook for just before execution happens (`on_notebook_start`, `on_cell_start`)
- hook for when execution completes (`on_notebook_complete`, `on_cell_complete`)
- hook for when execution has an error (`on_notebook_error`, `on_cell_error`)

Reply:
Added `on_notebook_error`, although I'm not sure if I did that part correctly. The only time I've personally seen the notebook fail is when there's a RuntimeError, so that's where I've placed the hook.

Reply:
fair enough - and feel free to push back if you think that this is "too many hooks all at once" 😅 on the one hand I like symmetry and intentional design for extension points. On the other hand, I also believe in not building things until somebody actively wants it to be built :-)  so if you think this is too much complexity, another option could be to just add comments for where you think the extra hooks could go, and wait to build them until a user explicitly asks for it.